python/samples/demos/guided_conversations/guided_conversation/utils/resources.py

```diff
@@ -0,0 +1,259 @@
+        quantity = round(quantity, 1)
+    unit = unit.value
+    return f"{quantity} {unit[:-1] if quantity == 1 else unit}"
+
+
+class GCResource:
+    """Resource constraints for the GuidedConversation agent. This class is used to keep track of the resource
+    constraints. If resource_constraint is None, then the agent can continue indefinitely. This also means
+    that no agenda will be created for the conversation.
+
+    Args:
+        resource_constraint (ResourceConstraint | None): The resource constraint for the conversation.
+        initial_seconds_per_turn (int): The initial number of seconds per turn. Defaults to 120 seconds.
+    """
+
+    def __init__(
+        self,
+        resource_constraint: ResourceConstraint | None,
+        initial_seconds_per_turn: int = 120,
+    ):
+        logger = logging.getLogger(__name__)
+        self.logger = logger
+        self.resource_constraint: ResourceConstraint | None = resource_constraint
+        self.initial_seconds_per_turn: int = initial_seconds_per_turn
+
+        self.turn_number: int = 0
+        self.remaining_units: float | None = None
+        self.elapsed_units: float | None = None
+
+        if resource_constraint is not None:
+            self.elapsed_units = 0
+            self.remaining_units = resource_constraint.quantity
+
+    def start_resource(self) -> None:
+        """To be called at the start of a conversation turn"""
+        if self.resource_constraint is not None and (
+            self.resource_constraint.unit == ResourceConstraintUnit.SECONDS
+            or self.resource_constraint.unit == ResourceConstraintUnit.MINUTES
+        ):
+            self.start_time = time.time()
+
+    def increment_resource(self) -> None:
+        """Increment the resource counter by one turn."""
+        if self.resource_constraint is not None:
+            if self.resource_constraint.unit == ResourceConstraintUnit.SECONDS:
+                self.elapsed_units += time.time() - self.start_time
+                self.remaining_units = self.resource_constraint.quantity - self.elapsed_units
+            elif self.resource_constraint.unit == ResourceConstraintUnit.MINUTES:
+                self.elapsed_units += (time.time() - self.start_time) / 60
+                self.remaining_units = self.resource_constraint.quantity - self.elapsed_units
+            elif self.resource_constraint.unit == ResourceConstraintUnit.TURNS:
+                self.elapsed_units += 1
+                self.remaining_units -= 1
+
+        self.turn_number += 1
+
+    def get_resource_mode(self) -> ResourceConstraintMode:
+        """Get the mode of the resource constraint.
+
+        Returns:
+            ResourceConstraintMode | None: The mode of the resource constraint, or None if there is no
+                resource constraint.
+        """
+        return self.resource_constraint.mode if self.resource_constraint is not None else None
+
+    def get_turn_number(self) -> int:
```

Comment:
do we need a get method for this? they can just `resource.turn_number`

Reply:
Resolved